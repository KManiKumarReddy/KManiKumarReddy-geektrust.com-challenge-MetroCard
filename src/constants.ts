export const PASSENGER_TYPES = {
  SENIOR_CITIZEN: 'SENIOR_CITIZEN',
  ADULT: 'ADULT',
  KID: 'KID',
};

export const STATION_NAMES = {
  AIRPORT: 'AIRPORT',
  CENTRAL: 'CENTRAL',
};

export const FARES = {
  [PASSENGER_TYPES.ADULT]: 200,
  [PASSENGER_TYPES.SENIOR_CITIZEN]: 100,
  [PASSENGER_TYPES.KID]: 50,
};

export const ERROR_CODES = {
  INSUFFICIENT_BALANCE: 'INSUFFICIENT_BALANCE',
};
