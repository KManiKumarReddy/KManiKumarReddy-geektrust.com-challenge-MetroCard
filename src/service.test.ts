import { PASSENGER_TYPES, STATION_NAMES } from './constants';
import MetroCardService from './service';

describe('MetroCardService', () => {
  it('should output right', () => {
    const logSpy = jest.spyOn(console, 'log');
    const service = new MetroCardService();

    service.registerCard('MC1', 400);
    service.registerCard('MC2', 100);
    service.registerCard('MC3', 50);
    service.checkIn(
      'MC1',
      PASSENGER_TYPES.SENIOR_CITIZEN,
      STATION_NAMES.AIRPORT
    );
    service.checkIn('MC2', PASSENGER_TYPES.KID, STATION_NAMES.AIRPORT);
    service.checkIn('MC3', PASSENGER_TYPES.ADULT, STATION_NAMES.CENTRAL);
    service.checkIn(
      'MC1',
      PASSENGER_TYPES.SENIOR_CITIZEN,
      STATION_NAMES.CENTRAL
    );
    service.checkIn('MC3', PASSENGER_TYPES.ADULT, STATION_NAMES.AIRPORT);
    service.checkIn('MC3', PASSENGER_TYPES.ADULT, STATION_NAMES.CENTRAL);

    service.printSummary();
    expect(logSpy).toHaveBeenCalledTimes(2);
    expect(logSpy.mock.calls[0][0]).toMatchSnapshot();
    expect(logSpy.mock.calls[1][0]).toMatchSnapshot();
  });

  it('should throw duplicate card error', () => {
    const service = new MetroCardService();
    ``;
    service.registerCard('MC1', 400);
    expect(() => service.registerCard('MC1', 100)).toThrow(
      'Duplicate card error: A card with same number already exists'
    );
  });

  it('should throw card not found error', () => {
    const service = new MetroCardService();
    service.registerCard('MC1', 400);
    expect(() =>
      service.checkIn('MC2', PASSENGER_TYPES.ADULT, STATION_NAMES.AIRPORT)
    ).toThrow("Invalid Input: Card MC2 doesn't exist");
  });

  it('should throw invalid station error', () => {
    const service = new MetroCardService();
    service.registerCard('MC1', 400);
    expect(() => service.checkIn('MC1', PASSENGER_TYPES.ADULT, 'TEST')).toThrow(
      "Invalid Input: Station TEST doesn't exist"
    );
  });
});
