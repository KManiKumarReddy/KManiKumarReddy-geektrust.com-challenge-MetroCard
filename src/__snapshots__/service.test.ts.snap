// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetroCardService should output right 1`] = `
"[1mTOTAL_COLLECTION[0m	CENTRAL	457	50
[1mPASSENGER_TYPE_SUMMARY[0m
ADULT	2
SENIOR_CITIZEN	1"
`;

exports[`MetroCardService should output right 2`] = `
"[1mTOTAL_COLLECTION[0m	AIRPORT	252	100
[1mPASSENGER_TYPE_SUMMARY[0m
ADULT	1
KID	1
SENIOR_CITIZEN	1"
`;
